File: src/main/java/com/lalkalol/mujahid/commands/impl/LeaveCommand.java
package com.lalkalol.mujahid.commands.impl;

import com.lalkalol.mujahid.commands.Command;
import com.lalkalol.mujahid.commands.CommandContext;
import com.lalkalol.mujahid.metrics.MetricsHolder;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData;

public class LeaveCommand implements Command {
    @Override
    public String name() {
        return "leave";
    }

    @Override
    public SlashCommandData data() {
        return Commands.slash("leave", "Leave the voice channel and clear the queue.");
    }

    @Override
    public void execute(CommandContext ctx) {
        if (ctx.getSelfVoiceChannel() == null) {
            ctx.replyEphemeral("I'm not connected to a voice channel.");
            return;
        }
        var existing = ctx.getLavalink().getExisting(ctx.getGuild().getIdLong());
        if (existing != null) {
            existing.getScheduler().stop();
        }
        ctx.getEvent().getJDA().getDirectAudioController().disconnect(ctx.getGuild());
        ctx.getLavalink().destroy(ctx.getGuild().getIdLong());
        MetricsHolder.get().recordVoiceSession("leave");
        MetricsHolder.get().setActivePlayers(ctx.getLavalink().activePlayerCount());
        ctx.reply("Left the voice channel. Bye!");
    }
}

File: src/main/kotlin/com/lalkalol/mujahid/commands/impl/LeaveCommand.kt
package com.lalkalol.mujahid.commands.impl

import com.lalkalol.mujahid.commands.Command
import com.lalkalol.mujahid.commands.CommandContext
import com.lalkalol.mujahid.metrics.MetricsHolder
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData

class LeaveCommand : Command {
    override fun name(): String = "leave"

    override fun data(): SlashCommandData =
        Commands.slash("leave", "Leave the voice channel and clear the queue.")

    override suspend fun execute(ctx: CommandContext) {
        if (ctx.getSelfVoiceChannel() == null) {
            ctx.replyEphemeral("I'm not connected to a voice channel.")
            return
        }
        ctx.getLavalink().getExisting(ctx.getGuild().idLong)?.scheduler?.stop()
        ctx.event.jda.directAudioController.disconnect(ctx.getGuild())
        ctx.getLavalink().destroy(ctx.getGuild().idLong)
        MetricsHolder.get().recordVoiceSession("leave")
        MetricsHolder.get().setActivePlayers(ctx.getLavalink().activePlayerCount())
        ctx.reply("Left the voice channel. Bye!")
    }
}
